trinity/protocol/common/handlers.py

```diff
@@ -0,0 +1,43 @@
+from abc import abstractmethod
+from typing import (
+    Dict,
+    Type,
+)
+
+from p2p.peer import BasePeer
+from p2p.protocol import _DecodedMsgType
+from p2p.service import BaseService
+
+from .managers import BaseRequestManager
+from .requests import BaseRequest
+
+
+class BaseRequestResponseHandler(BaseService):
+    @property
+    @abstractmethod
+    def _managers(self) -> Dict[str, Type[BaseRequestManager[BasePeer, BaseRequest, _DecodedMsgType, _DecodedMsgType]]]:
+        pass
+
+    def __init__(self, peer: BasePeer) -> None:
+        self._peer = peer
+
+        super().__init__(peer.cancel_token)
+        for attr, request_manager_cls in self._managers.items():
+            if hasattr(self, attr):
+                raise AttributeError(
+                    "Unable to set manager on attribute `{0}` which is already "
+                    "present on the class: {1}".format(attr, getattr(self, attr))
+                )
+            manager = request_manager_cls(self._peer, self.cancel_token)
+            setattr(self, attr, manager)
+
+    async def _run(self) -> None:
+        for attr in self._managers.keys():
+            manager = getattr(self, attr)
+            self.run_child_service(manager)
```

Comment:
So we're talking about `num_peers * num_commands` services running, right? Not sure how many commands there will be, but let's say eventually 50. If someone wanted to experiment with 200 peers, we're talking 10,000 different services running. Not sure when overhead of multiple tasks starts to overwhelm asyncio, but I wouldn't be surprised if it was an issue at 10k.

Reply:
Formula is a little off but it doesn't discount your concern entirely.

`num_peers * `num_command_pairs`

Currently I think the `eth` protocol has 4-5 command pairs and the LES protocol has maybe 8-10.  So only request/response command pairs. That at least pushes the overhead of this down by a decent bit, and for now, I really like how clean it is for this to be self contained.  We can reduce the overhead by moving the subscription from the `Manager` classes down to the `Handler` classes and having the handler dispatch to each of the mangers which moves it from `O(n)` to `O(1)` per peer.